File: weather_collector/processors/sea_breeze.py
"""
Sea breeze detection for coastal locations
"""
from datetime import datetime


def detect_sea_breeze(weather_data):
    """
    Detect sea breeze likelihood based on land/water temp differential,
    wind speed, direction, and time of day.
    """
    current = weather_data.get("current", {})
    buoy = weather_data.get("buoy_44013", {})
    
    pws_temp = current.get("temperature")
    water_temp = buoy.get("water_temp_f")
    wind_speed = current.get("wind_speed")
    wind_dir = current.get("wind_direction")
    
    now = datetime.now()
    hour = now.hour
    
    result = {
        "active": False,
        "likelihood": 0,
        "reason": ""
    }
    
    # Check data availability
    if None in [pws_temp, water_temp, wind_speed, wind_dir]:
        result["reason"] = "Insufficient data"
        weather_data["sea_breeze"] = result
        return
    
    temp_diff = pws_temp - water_temp
    scores = {}
    
    # Temperature differential
    if temp_diff < 0:
        scores["temp"] = 0
    elif temp_diff < 3:
        scores["temp"] = 20
    elif temp_diff < 5:
        scores["temp"] = 40
    elif temp_diff < 8:
        scores["temp"] = 70
    elif temp_diff < 12:
        scores["temp"] = 90
    else:
        scores["temp"] = 100
    
    # Wind speed
    if wind_speed > 15:
        scores["wind_speed"] = 0
    elif wind_speed > 12:
        scores["wind_speed"] = 30
    elif wind_speed > 10:
        scores["wind_speed"] = 60
    elif wind_speed > 7:
        scores["wind_speed"] = 90
    else:
        scores["wind_speed"] = 100
    
    # Wind direction (270-330° from Salem Harbor)
    onshore_min, onshore_max = 260, 340
    if onshore_min <= wind_dir <= onshore_max:
        scores["direction"] = 100
    elif 240 <= wind_dir < onshore_min or onshore_max < wind_dir <= 360:
        scores["direction"] = 50
    elif 0 <= wind_dir <= 20:
        scores["direction"] = 50
    else:
        scores["direction"] = 0
    
    # Time of day
    if 10 <= hour < 18:
        scores["time"] = 100
    elif 8 <= hour < 10 or 18 <= hour < 20:
        scores["time"] = 60
    else:
        scores["time"] = 0
    
    # Overall likelihood
    likelihood = int(
        scores["temp"] * 0.40 +
        scores["direction"] * 0.30 +
        scores["wind_speed"] * 0.20 +
        scores["time"] * 0.10
    )
    
    result["likelihood"] = likelihood
    
    # Determine if active
    if likelihood >= 60:
        result["active"] = True
        result["reason"] = f"Sea breeze active: Δ{temp_diff:+.1f}°F, {wind_speed:.0f} mph from {int(wind_dir)}°"
    elif likelihood >= 40:
        result["reason"] = f"Possible sea breeze: Δ{temp_diff:+.1f}°F, {wind_speed:.0f} mph from {int(wind_dir)}°"
    elif scores["temp"] < 40:
        result["reason"] = f"Land/water Δ too small ({temp_diff:+.1f}°F)"
    elif scores["direction"] < 50:
        result["reason"] = f"Wind not onshore ({int(wind_dir)}°, need 270-330°)"
    elif scores["wind_speed"] < 50:
        result["reason"] = f"Winds too strong ({wind_speed:.0f} mph)"
    elif scores["time"] == 0:
        result["reason"] = f"Wrong time of day ({hour:02d}:00, need 10am-6pm)"
    else:
        result["reason"] = "Conditions not met"
    
    result["scores"] = scores
    weather_data["sea_breeze"] = result

File: js/seabreeze.js
// seabreeze.js — Sea breeze forecast card

function renderSeaBreezeDetail(data) {
  const sb = data.sea_breeze || {};
  const el = document.getElementById("seaBreezeDetail");
  if (!el) return;

  if (sb.likelihood === undefined || sb.likelihood === null) {
    el.innerHTML = `<div style="text-align:center;color:rgba(255,255,255,0.5);padding:20px;">No sea breeze data available</div>`;
    return;
  }

  const scores = sb.scores || {};
  const hyp = data.hyperlocal || {};
  const buoy = data.buoy_44013 || {};
  const cur = data.current || {};

  let statusColor, statusText;
  if (sb.active) {
    statusColor = "rgba(100,200,120,0.95)";
    statusText = "Active";
  } else if (sb.likelihood >= 40) {
    statusColor = "rgba(220,200,60,0.85)";
    statusText = "Possible";
  } else {
    statusColor = "rgba(150,150,150,0.6)";
    statusText = "Unlikely";
  }

  // Build sea breeze headline
  let sbHeadline;
  if (sb.active) {
    sbHeadline = `Sea breeze active — offshore flow replaced by onshore`;
  } else if (sb.likelihood >= 60) {
    sbHeadline = `Sea breeze likely this afternoon (${sb.likelihood}%)`;
  } else if (sb.likelihood >= 35) {
    sbHeadline = `Sea breeze possible (${sb.likelihood}%) — conditions marginal`;
  } else {
    const windDir = data.current?.wind_direction;
    const compass = windDir != null ? toCompass(windDir) : null;
    sbHeadline = compass
      ? `No sea breeze — ${compass} wind dominates (${sb.likelihood}%)`
      : `No sea breeze — conditions unfavorable (${sb.likelihood}%)`;
  }
  const sbHeadlineColor = sb.active ? "rgba(100,200,120,0.95)" : sb.likelihood >= 40 ? "rgba(220,200,60,0.85)" : "rgba(150,150,150,0.7)";

  const html = `
    <div style="font-size:0.95rem;font-weight:600;color:${sbHeadlineColor};margin-bottom:16px;padding:10px 14px;background:rgba(255,255,255,0.04);border-radius:0;border-left:4px solid ${sbHeadlineColor};">${sbHeadline}</div>
    <div style="text-align:center;margin-bottom:20px;">
      <div style="font-size:2.5rem;color:${statusColor};margin-bottom:8px;">${sb.likelihood}%</div>
      <div style="font-size:1.1rem;opacity:0.9;">${statusText}</div>
      <div style="font-size:0.9rem;opacity:0.7;margin-top:4px;">${sb.reason}</div>
    </div>

    <div style="display:grid;grid-template-columns:repeat(auto-fit,minmax(200px,1fr));gap:16px;margin-bottom:20px;">
      <div>
        <div style="opacity:0.7;font-size:0.85rem;margin-bottom:4px;">Temp Differential</div>
        <div style="font-size:1.3rem;">${scores.temp || 0}%</div>
        <div style="opacity:0.6;font-size:0.8rem;margin-top:2px;">Land: ${(hyp.corrected_temp ?? cur.temperature)?.toFixed(1) || "--"}°F | Water: ${buoy.water_temp_f?.toFixed(1) || "--"}°F</div>
      </div>
      <div>
        <div style="opacity:0.7;font-size:0.85rem;margin-bottom:4px;">Wind Speed</div>
        <div style="font-size:1.3rem;">${scores.wind_speed || 0}%</div>
        <div style="opacity:0.6;font-size:0.8rem;margin-top:2px;">${(hyp.corrected_wind_speed ?? cur.wind_speed)?.toFixed(1) || "--"} mph</div>
      </div>
      <div>
        <div style="opacity:0.7;font-size:0.85rem;margin-bottom:4px;">Direction</div>
        <div style="font-size:1.3rem;">${scores.direction || 0}%</div>
        <div style="opacity:0.6;font-size:0.8rem;margin-top:2px;">${cur.wind_direction ? Math.round(cur.wind_direction) + "° " + toCompass(cur.wind_direction) : "--"}</div>
      </div>
      <div>
        <div style="opacity:0.7;font-size:0.85rem;margin-bottom:4px;">Time of Day</div>
        <div style="font-size:1.3rem;">${scores.time || 0}%</div>
        <div style="opacity:0.6;font-size:0.8rem;margin-top:2px;">${new Date().toLocaleTimeString("en-US", {hour: "numeric", minute: "2-digit"})}</div>
      </div>
    </div>

    <div style="opacity:0.7;font-size:0.85rem;line-height:1.5;">
      <strong>Calculation:</strong> Likelihood = (Temp × 40%) + (Direction × 30%) + (Wind Speed × 20%) + (Time × 10%)<br>
      = (${scores.temp || 0} × 0.4) + (${scores.direction || 0} × 0.3) + (${scores.wind_speed || 0} × 0.2) + (${scores.time || 0} × 0.1) = ${sb.likelihood}%
    </div>
  `;

  el.innerHTML = html;
  
  // Update collapsed preview with new structure
  const seaBreezeCollapsedEl = document.getElementById("seaBreezeCollapsed");
  const seaBreezeLabelEl = document.getElementById("seaBreezeLabel");
  const seaBreezeProbCollapsedEl = document.getElementById("seaBreezeProbCollapsed");
  if (seaBreezeCollapsedEl) {
    seaBreezeCollapsedEl.innerHTML = `${sb.likelihood}<span style="font-size:1.8rem;opacity:0.6;">%</span>`;
  }
  if (seaBreezeLabelEl) seaBreezeLabelEl.textContent = statusText;
  if (seaBreezeProbCollapsedEl) {
    let timeText;
    if (sb.active) {
      timeText = "Active now";
    } else if (sb.likelihood >= 40) {
      timeText = "This afternoon";
    } else {
      const wd = data.current?.wind_direction;
      timeText = wd != null ? `Wind from ${toCompass(wd)}` : "Unfavorable";
    }
    seaBreezeProbCollapsedEl.textContent = timeText;
  }
  
  // Apply gradient class based on likelihood
  const seaBreezeCard = document.querySelector('[data-collapse-key="sea_breeze_detail"]');
  if (seaBreezeCard) {
    seaBreezeCard.classList.remove('tile-seabreeze-unlikely', 'tile-seabreeze-possible', 'tile-seabreeze-likely');
    if (sb.likelihood < 30) seaBreezeCard.classList.add('tile-seabreeze-unlikely');
    else if (sb.likelihood < 60) seaBreezeCard.classList.add('tile-seabreeze-possible');
    else seaBreezeCard.classList.add('tile-seabreeze-likely');
  }
}
